proposals/interface-types/working-notes/record.md

```diff
@@ -0,0 +1,73 @@
+
+
+### Records
+
+Records are collections of named and typed fields. Records can be defined in as
+Interface Types; for example this definition encodes a familiar notion of a
+credit card:
+
+```wat
+(@interface datatype @cc 
```

Comment:
For symmetry with how we have `(func $f)`/`(@interface func $f)`, how about `(type $T)`/`(@interface type $T)` (instead of `datatype @T`)?

Reply:
Am open to adjustment. However, when defining types there are two basic scenarios: defining a new algebraic type and type aliasing. These need to be distinguished. (Examples of 'type aliasing' include type imports)

Reply:
@fgmccabe, the only reason to distinguish them would be if you wanted nominal typing for algebraic types. But there is little reason to want that in Wasm interfaces. In fact, there is good reason not to want that because it makes everything more complicated and aversely affects modularity.

Reply:
On second thoughts, I think we need to keep the @ character.
If you have a param specification for an interface function; you need to disambiguate:

(param $cc (resource $connection))

could mean a param block with two types that happen to be by reference or a single parameter.

On the other hand

(param @cc (resource @connection))

is unambiguous (a param block of two types)

Reply:
Ah, interesting question.  If we look at the [GC proposal](https://github.com/WebAssembly/gc/blob/master/proposals/gc/Overview.md), which similarly introduces type definitions and uses, we can see that, in most cases, a type use occurs as part of a reference type, so you can unambiguously have `(ref $T)` for some typedef `$T`.  However, there is one exception: when the field of a `struct` or element of an `array` is stored inline, not by reference; for this the explainer uses `(type $T)` which is presumably chosen to be symmetric with the MVP text format, where you can write:
```wasm
(module
  (type $F (func (param i32) (result i32)))
  (func $f (type $F) ...)
)
```
as the expanded form of the syntactic sugar:
```wasm
(module
  (func $f (param i32) (result i32) ...)
)
```
Thus, I think your final example would be most-symmetrically written:
```wasm
(param (type $cc) (resource $cc))
```